statsmodels/stats/power.py

```diff
@@ -172,39 +172,49 @@ def normal_sample_size_one_tail(diff, power, alpha, std_null=1.,
     """explicit sample size computation if only one tail is relevant
 
     The sample size is based on the power in one tail assuming that the
-    alternative is in the tail where the test has power that increases
-    with sample size.
-    Use alpha/2 to compute the one tail approximation to the two-sided
-    test, i.e. consider only one tail of two-sided test.
+    alternative is in the tail where the test has power that increases with
+    sample size. Use alpha/2 to compute the one tail approximation to the
+    two-sided test, i.e. consider only one tail of two-sided test.
 
     Parameters
     ----------
     diff : float
         difference in the estimated means or statistics under the alternative.
-    power : float in interval (0,1)
-        power of the test, e.g. 0.8, is one minus the probability of a type II error.
-        Power is the probability that the test correctly rejects the Null
-        Hypothesis if the Alternative Hypothesis is true.
+    power : float in interval [alpha,1)
+        power of the test, e.g. 0.8, is one minus the probability of a type II
+        error. Power is the probability that the test correctly rejects the
+        Null Hypothesis if the Alternative Hypothesis is true.
     alpha : float in interval (0,1)
-        significance level, e.g. 0.05, is the probability of a type I
-        error, that is wrong rejections if the Null Hypothesis is true.
+        significance level, e.g. 0.05, is the probability of a type I error,
+        that is wrong rejections if the Null Hypothesis is true.
     std_null : float
         standard deviation under the Null hypothesis without division by
         sqrt(nobs)
-    std_alternative : float
-        standard deviation under the Alternative hypothesis without division
-        by sqrt(nobs)
+    std_alternative : float (optional)
+        standard deviation under the Alternative hypothesis without division by
+        sqrt(nobs). Defaults to None. If None, ``std_alternative`` is set to the
+        value of ``std_null``.
 
     Returns
     -------
     nobs : float
         sample size to achieve the desired power
 
     """
+    if std_alternative is None:
+        std_alternative = std_null
 
     crit_power = stats.norm.isf(power)
     crit = stats.norm.isf(alpha)
     n1 = ((crit * std_null - crit_power * std_alternative) / diff)**2
+
+    # Enforce condition that power >= alpha; return np.nan for elements not
+    # satisfying this condition. np.ones_like used to satisfy broadcastability
+    # constraints and to yield a Boolean array with the same shape as n1.
+    alpha_minus_power = (
+        alpha * np.ones_like(std_null) - power * np.ones_like(std_alternative)
+        ) / np.ones_like(diff)
```

Comment:
diff could be negative, then it could be -/- = +

Reply:
I'm not sure I understand your comment; the denominator in this expression is an array of all ones with the same shape as `diff` -- the values in `diff` do not change the output of `numpy.ones_like`. The point of using `np.ones_like` is to build `numpy.NDArray[numpy.float64]`s of all ones to satisfy array broadcasting requirements due to the vectorization feature you've requested I preserve.

Reply:
sorry, I misread

However, we just need (line 210)
`n1 = (np.maximum(crit * std_null - crit_power * std_alternative, 0) / diff)**2`

alpha < power is not the appropriate condition for the term to be nonegative if std_null != std_alternative

Reply:
If `crit`, `std_null`, `crit_power`, `std_alternative`, and `diff` are all real, `n1` can never be negative because it is the (elementwise) square of real quantities, so the proposed `numpy.maximum` operation is unnecessary, and will not fix the problem regarding `power <= alpha` for a one-tailed test. The problem is one of well-posedness, not one of algebra or the sign of different quantities.

If we assume that `std_null == std_alternative`, then the situation for a one-tailed test resembles that of this graphic (https://en.wikipedia.org/wiki/Power_of_a_test#/media/File:Statistical_test,_significance_level,_power.png), if we ignore the blue-shaded region in the lower-tail of the blue curve. When `power <= alpha`, the mean of the red curve is less than or equal to the mean of the blue curve, and you cannot have a "false negative" in that situation -- for a one-tailed test, it simply does not make sense because the specific alternative being used in the power analysis will be an alternative consistent with the null hypothesis.

For a two-tailed test, the situation is different because the specific alternative being used in the power analysis will jump from the lower tail to the upper tail, or vice versa.

Reply:
If I remember the derivation for the `std_null != std_alternative` case correctly, the test can be posed in terms of the difference of the means of the "null" and "alternative", and a similar sort of situation applies for a one-tailed test. However, the formula I got when I did that derivation was that of page 490 in the screenshots posted in https://github.com/statsmodels/statsmodels/pull/8507#issuecomment-1312146226, not the formula used in the implementation of `normal_sample_size_one_tail`.

Reply:
For proportions, the formulas on page 452 of the screenshots posted in https://github.com/statsmodels/statsmodels/pull/8507#issuecomment-1312146226 are different than for non-proportions under the assumption that the variance of a proportion with value `p` is `n * p * (1 - p)`, so the `Z` statistic in that case is not a standard normal variable with zero mean and unit variance. I would have to try to re-derive the result myself to understand better what is happening.

Reply:
At this point, I'd be willing to just drop all of the changes related to `alpha` and `power`. My impression is that we disagree about the cause of the problem when `power <= alpha` and when it occurs. The changes regarding multiplication by `None` are less contentious, and we agree on those, so it might be better to land those changes.

re: the `power <= alpha` issue, if that's something that you agree exists and should be fixed, I would be interested in continuing those discussions if we can come to some sort of agreement about what to do, but I fear we are at an impasse.

re: the implementation, I'd also be happy to use `statsmodels.stats.power.normal_sample_size_one_tail` if I understood the assumptions under which it is valid, with references to back that up. Otherwise, I can't tell whether it's appropriate for my application, and can't make a business case to my management that we should use and support this library.

Whatever your decision, I support whatever you think is best. Software maintenance is a lot of work, and it's often under-recognized, particularly when it comes to numerical software, so I want to be conscious of that and want to avoid being a drain on your time or mental health.